internal/langserver/handlers/initialize.go

```diff
@@ -188,8 +105,150 @@ func (svc *service) Initialize(ctx context.Context, params lsp.InitializeParams)
 		})
 	}
 
+	if params.RootURI == "" {
+		svc.singleFileMode = true
+		properties["root_uri"] = "file"
+		// return serverCaps, fmt.Errorf("Editing a single file is not yet supported." +
+		// 	" Please open a directory.")
+		jrpc2.ServerFromContext(ctx).Notify(ctx, "window/showMessage", &lsp.ShowMessageParams{
+			Type:    lsp.Warning,
+			Message: "Some of VS Code's capabilities are reduced when editing a single file, but you can still open text files and edit them. Please open a directory for full functionality.",
+		})
+	} else {
+		if !uri.IsURIValid(string(params.RootURI)) {
+			properties["root_uri"] = "invalid"
+			return serverCaps, fmt.Errorf("URI %q is not valid", params.RootURI)
+		}
+		svc.setWalker(ctx, params, *out.Options)
+		if err != nil {
+			return serverCaps, err
+		}
+	}
+
+	// Walker runs asynchronously so we're intentionally *not*
+	// passing the request context here
+	walkerCtx := context.Background()
+	err = svc.walker.StartWalking(walkerCtx)
+	if err != nil {
+		return serverCaps, err
+	}
+
+	return serverCaps, nil
+}
+
+func setupTelemetry(expClientCaps lsp.ExpClientCapabilities, svc *service, ctx context.Context, properties map[string]interface{}) {
+	if tv, ok := expClientCaps.TelemetryVersion(); ok {
+		svc.logger.Printf("enabling telemetry (version: %d)", tv)
+		err := svc.setupTelemetry(tv, svc.server)
+		if err != nil {
+			svc.logger.Printf("failed to setup telemetry: %s", err)
+		}
+		svc.logger.Printf("telemetry enabled (version: %d)", tv)
+	}
+	defer svc.telemetry.SendEvent(ctx, "initialize", properties)
+}
+
+func getProperties(out *settings.DecodedOptions) map[string]interface{} {
```

Comment:
```suggestion
func getTelemetryProperties(out *settings.DecodedOptions) map[string]interface{} {
```

Reply:
I understand removing the `get` part, but do not understand the `telemetry` part. This builds up the struct for the settings sent by the client, not the telemetry. Though I admit I don't know what other name to name it....